DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1

```diff
@@ -1638,4 +1639,401 @@ function Assert-FileSignatureValid
     }
 }
 
+<#
+    .SYNOPSIS
+        Starts and waits for a process.
+
+    .DESCRIPTION
+        Allows mocking and testing of process arguments.
+
+    .PARAMETER Process
+        The System.Diagnositics.Process object to start.
+
+    .PARAMETER LogStream
+        Redirect STDOUT and STDERR output.
+#>
+function Invoke-Process
+{
+    [CmdletBinding()]
+    [OutputType([System.Diagnostics.Process])]
+    param (
+        [Parameter(Mandatory)]
+        [System.Diagnostics.Process]
+        $Process,
+
+        [Parameter()]
+        [System.Boolean]
+        $LogStream
+    )
+
+    $Process.Start() | Out-Null
+
+    if ($LogStream)
+    {
+        $Process.BeginOutputReadLine()
+        $Process.BeginErrorReadLine()
+    }
+
+    $Process.WaitForExit()
+    return $Process
+}
+
+<#
+    .SYNOPSIS
+        Runs a process as the specified user via PInvoke.
+
+    .PARAMETER CommandLine
+        The command line (including arguments) of the process to start.
+
+    .PARAMETER Credential
+        The user credential to start the process as.
+#>
+function Invoke-PInvoke
+{
+    [CmdletBinding()]
+    [OutputType([System.Int32])]
+    param
+    (
+        [Parameter(Mandatory = $true)]
+        [String]
+        $CommandLine,
+
+        [Parameter(Mandatory)]
+        [PSCredential]
+        [System.Management.Automation.CredentialAttribute()]
+        $Credential
+    )
+
+    Register-PInvoke
+    [System.Int32] $exitCode = 0
+
+    [Source.NativeMethods]::CreateProcessAsUser($CommandLine, `
+        $Credential.GetNetworkCredential().Domain, `
+        $Credential.GetNetworkCredential().UserName, `
+        $Credential.GetNetworkCredential().Password, `
+        [ref] $exitCode
+    )
+
+    return $exitCode;
+}
+
+<#
+    .SYNOPSIS
+        Registers PInvoke to run a process as a user.
+#>
+function Register-PInvoke
+{
+    $programSource = @'
+        using System;
```

Comment:
Did you get this code from another resource?  If so, please add a comment about where you got it from, please?